cmd/specdiff/differ.go

```diff
@@ -0,0 +1,283 @@
+	if s.isEmpty() {
+		return "", false
+	}
+	index := len(*s) - 1
+	element := (*s)[index]
+	*s = (*s)[:index]
+	return element, true
+}
+
+// GetDiff takes two yaml or json Open API 3 Specs and diffs them.
+func GetDiff(base, revision []byte) (*rpc.Diff, error) {
+	loader := openapi3.NewLoader()
+	loader.IsExternalRefsAllowed = true
+	baseSpec, err := loader.LoadFromData(base)
+	if err != nil {
+		err := fmt.Errorf("failed to load base spec from %q with %v", base, err)
+		return nil, err
+	}
+	revisionSpec, err := loader.LoadFromData(revision)
+	if err != nil {
+		err := fmt.Errorf("failed to load revision spec from %q with %v", revision, err)
+		return nil, err
+	}
+	diffReport, err := diff.Get(&diff.Config{
+		ExcludeExamples:    false,
+		ExcludeDescription: false,
+	}, baseSpec, revisionSpec)
+	if err != nil {
+		err := fmt.Errorf("diff failed with %v", err)
+		return nil, err
+	}
+	return getChanges(diffReport)
+}
+
+func addToDiffProto(diffProto *rpc.Diff, changePath *change) {
+	fieldName := changePath.fieldPath.String()
+	switch changePath.changeType {
+	case "added":
+		diffProto.Added = append(diffProto.Added, fieldName)
+	case "deleted":
+		diffProto.Deleted = append(diffProto.Deleted, fieldName)
+	}
+}
+
+// getChanges creates a protodif report from a diff.Diff struct.
+func getChanges(diff *diff.Diff) (*rpc.Diff, error) {
+	diffProto := &rpc.Diff{
+		Added:        []string{},
+		Deleted:      []string{},
+		Modification: make(map[string]*rpc.Diff_ValueChange),
+	}
+	change := &change{
+		fieldPath:  stack{},
+		changeType: "",
+	}
+	diffNode := reflect.ValueOf(diff)
+	err := searchNode(diffNode, diffProto, change)
+	return diffProto, err
+}
+
+func searchNode(value reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	// Some values might be pointers, so we should dereference before continuing.
+	value = reflect.Indirect(value)
+
+	// Invalid values aren't relevant to the diff.
+	if !value.IsValid() {
+		return nil
+	}
+
+	switch value.Kind() {
+	case reflect.Map:
+		return searchMapType(value, diffProto, changePath)
+	case reflect.Array, reflect.Slice:
+		return searchArrayAndSliceType(value, diffProto, changePath)
+	case reflect.Struct:
+		return searchStructType(value, diffProto, changePath)
+	case reflect.Float64, reflect.String, reflect.Bool, reflect.Int:
+		valueString := getAtomicType(value)
+		changePath.fieldPath.push(valueString)
+		addToDiffProto(diffProto, changePath)
+		changePath.fieldPath.pop()
+		return nil
+	default:
+		return fmt.Errorf("field %q has unknown type %s with value %v", changePath.fieldPath, value.Type(), value)
+	}
+}
+
+func searchMapType(mapNode reflect.Value, diffProto *rpc.Diff, changePath *change) error {
+	if mapNode.Kind() != reflect.Map {
+		panic("searchMapType called with invalid type")
+	}
+
+	for _, childNodeKey := range mapNode.MapKeys() {
+		childNode := mapNode.MapIndex(childNodeKey)
+		if childNode.IsZero() {
+			continue
+		}
+
+		if isAtomicType(childNodeKey) {
```

Comment:
Since both of these checks are looking at the type of the key, it might make more sense to do all the checking here if it works out nicely.
```
	switch childNodeKey.Interface().(type) {
	case float64, string, bool, int:
		// Handle scalar type.
	case diff.Endpoint:
		// Handle endpoint.
	default:
		// Return error.
	}
```